starlette/datastructures.py

```diff
@@ -290,6 +291,95 @@ def __repr__(self) -> str:
         return f"{self.__class__.__name__}({repr(items)})"
 
 
+class MultiDict(ImmutableMultiDict):
+    def __setitem__(self, key: typing.Any, value: typing.Any) -> None:
+        self.setlist(key, [value])
+        self._dict[key] = value
+
+    def __delitem__(self, key: typing.Any) -> None:
+        self._list = list(filter(lambda item: item[0] != key, self._list))
+        del self._dict[key]
+
+    def pop(self, key: typing.Any) -> typing.Any:
+        list_ = list(filter(lambda item: item[0] == key, self._list))
+        poped = self._dict.pop(key)
+        self.setlist(key, [v for _, v in list_[:-1]])
+        return poped
+
+    def popitem(self, key: typing.Any) -> typing.Tuple:
+        return key, self.pop(key)
+
+    def poplist(self, key: typing.Any) -> typing.List:
+        poped_list = list(filter(lambda item: item[0] == key, self._list))
+        self._list = list(filter(lambda item: item[0] != key, self._list))
+        self._dict.pop(key)
+        return [v for _, v in poped_list]
+
+    def clear(self) -> None:
+        self._dict.clear()
+        self._list.clear()
+
+    def setdefault(self, key: typing.Any, default: typing.Any = None) -> typing.Any:
+        if key not in self:
+            self._dict[key] = default
+            self._list.append((key, default))
+
+        return self[key]
+
+    def setlist(self, key: typing.Any, values: typing.List) -> None:
+        filter_obj = filter(lambda item: item[0] != key, self._list)
+        if not values:
+            values.append(None)
+        self._list = [*filter_obj, *((key, value) for value in values)]
+        self._dict[key] = self._list[-1][1]
+        # Case list is empty
+
+    def setlistdefault(
+        self, key: typing.Any, default_list: typing.List = None
+    ) -> typing.List:
+        if key not in self:
+            if default_list is None:
+                default_list = [None]  # or just []
```

Comment:
Just `[]` please. 😄

Reply:
Also let's just call it `default`

Reply:
if an empty was provided should we remove the item from dict or set the value in dict to None or raise an error ?

Reply:
Let's just omit this method entirely. I can't think of obvious enough use cases. It'd be reasonable to push users towards `if key not in q: q.setlist(values)`, which I think is more obvious.

Reply:
if values was empty, what should we do ?

```python
q = MultiDict([("a", "123")])
q.setlist("a", []) # ?
q.setlist("b", []) # ?
```

Reply:
Delete it from the dict.

Reply:
Alright !